docs: Proposal for a TargetNodeDirectory in Tractus-X. (#1556)
eclipse-tractusx/tractusx-edc: docs/development/decision-records/2024-09-16-tx-target-node-directory-proposal/README.md

# Proposal for a TargetNodeDirectory in Tractus-X

## Decision

A new service that will contain Connectors' URL's of each partner a member wants the offers from, acting as the **TargetNodeDirectory for the partner's FederatedCatalog**.

## Rationale

While considering new interventions in the Federated Catalog, this decision aims to set the TargetNodeDirectory.

> From the [documentation](https://eclipse-edc.github.io/docs/#/submodule/FederatedCatalog/docs/developer/architecture/federated-catalog.architecture)

Reviewer: this link is not working, maybe it's better to point directly to the readme in the repo:
https://github.com/eclipse-edc/FederatedCatalog/blob/e733355c6991ff633ee009bd5f35ced61e941da6/docs/developer/architecture/federated-catalog.architecture.md

Author: unsure why, but when created this doc that link pointed to correct documentation (that was the same as the readme.md file) and now it does not.
changed in [here](https://github.com/eclipse-tractusx/tractusx-edc/pull/1556/commits/a9cdd06d866b53b2ad56fbd0c92f0a67af751ae2).

Reviewer: that's because upstream we just switched to the new docs site (https://eclipse-edc.github.io), but FC documentation still needs to be moved

> The Federated Catalog requires a list of Target Catalog Nodes, so it knows which endpoints to crawl. This list is provided by the TargetNodeDirectory. During the preparation of a crawl run, the ExecutionManager queries that directory and obtains the list of TCNs.

> So having a service containing the data of Connectors' URL's that a certain partner wants and allow them to host it. Users are able to input (using the API of new service) the Connectors' URL's of the connectors they want the catalogs from.

Reviewer: what do we mean here with "service"? a class? a runtime? an endpoint? 
could we give it a name as well?

Author: the idea would be to have a runnable jar. a service able to run independently.
following a suggestion, maybe TargetNodeDirectoryService. 
Changed in [here](https://github.com/eclipse-tractusx/tractusx-edc/pull/1556/commits/a9cdd06d866b53b2ad56fbd0c92f0a67af751ae2).

> This solution allows the member to choose precisely the Target Catalog Nodes that interests them, resulting in reduced network calls and latency. Additionally, each member has control on how to host and manage this new service. Service changes do not affect other parties (unless contract changes) and is able to be independently scalable.

Reviewer: ```suggestion
This solution allows the member to choose precisely the Target Catalog Nodes that interests them, resulting in reduced network calls and latency. Additionally, each member has control on how to host and manage this new service. Service changes do not affect other parties (unless contract changes) and it can be scaled be independently.
```

Author: Changed in [here](https://github.com/eclipse-tractusx/tractusx-edc/pull/1556/commits/a9cdd06d866b53b2ad56fbd0c92f0a67af751ae2).

Other solution was also considered

- File in a S3 bucket (or different cloud provider's solution)
    - This solution was discarded due to one file for all instead of each partner having the data that respectively needs does not match the requirement and this solution would lock the usage of a proprietary tool (cloud provider) being harder to sustain in the long run.

## Approach

> The user is able to obtain the Connectors' URL's (through the Discovery Service, as an example) and store them in the new service through its API. The API will allow to save a list of Connectors' URL's in bulk and the service is responsible to store that (in memory or in database). These can later be retrieved and crawled by the Federated Catalog.

Reviewer: is this operation meant to be done manually?

Author: Yes, with the goal of the removing the Discovery Service dependency from this TargetNodeDirectoryService the responsibility of obtaining and saving the the Connectors' URL's lies with the user, as sugggested [here](https://github.com/eclipse-tractusx/tractusx-edc/pull/1556#issuecomment-2376486527).

# Proposal for a TargetNodeDirectory in Tractus-X

## Decision

A new service that will have a db/cache containing the BPNL's and Connectors' URL's of each partner a member wants the offers from. Simply, **this new service works as the TargetNodeDirectory for the partner's FederatedCatalog**.

## Rationale

While considering new interventions in the Federated Catalog, this discussion aims to propose a solution for the TargetNodeDirectory.
From the [documentation](https://eclipse-edc.github.io/docs/#/submodule/FederatedCatalog/docs/developer/architecture/federated-catalog.architecture)
> The Federated Catalog requires a list of Target Catalog Nodes, so it knows which endpoints to crawl. This list is provided by the TargetNodeDirectory. During the preparation of a crawl run, the ExecutionManager queries that directory and obtains the list of TCNs.

So having a service containing the data of desired BPNL's and respective Connectors' URL's that a certain partner wants, but here the partner itself has the service in their scope and users are able to input (in the API of new service) the Participants.

> This solution allows the member to choose the relevant data that interests them (offers from specific participants) enabling the reduction of storage cost and less network calls. Additionally, each member has control on how to host and manage this new service, service changes do not affect other parties (unless contract changes) and is able to be independently scalable.

Reviewer 2: ```suggestion
This solution allows the member to choose precisely the Target Catalog Nodes that interests them reducing storage cost and network calls. Additionally, each member has control on how to host and manage this new service, service changes do not affect other parties (unless contract changes) and is able to be independently scalable.
```

Author: Yes, nice touch, added [here](https://github.com/eclipse-tractusx/tractusx-edc/pull/1556/commits/fae5e8bafbc033e16bcfce259fc95cb2ad7b06af).

- Centralized service with all participants data
    - Solution was not considered due to presenting a single point of failure, the partner still needs to have a way to dynamically store on desired partners ids' and each partner must have the BPNL or the Connectors' URL's beforehand.

## Approach

Using the Discovery Service, the FC is able to request the connetors' url's that it requires, providing the BPNL's. Since the Discovery Service has already built in this feature, the work of this flow is already relatively well established.
However, the main question here arises: what if a new partner joins the space? The existing partners that want to obtain the offers of the new partner do not know of its existence. So an additional mechanism is needed. The default implementation is a static file, however a dynamic approach would be preferable (no downtime expected).

Even with a new service registering the Discovery Service (basically making it available in the space to other partners) the existing partners that may want to retrieve the new partners' offers must establish the need to retrieve them by defining the new BPNL somewhere.

> A service containing the data of desired BPNL's and respective Connectors' URL's that a certain partner wants, but here the partner itself has the service in their scope and users are able to input (in the API of new service) the Participants. Simply, the new service would have a db/cache containing the BPNL's and Connectors' URL's of each partner this member wants the offers from. Again, **this new service works as the TargetNodeDirectory for the partner's FederatedCatalog**.

Reviewer 2: If you feel the need to put something in simpler terms, why do you keep both forms and not only the simpler one?

Author: Yeah. Removed duplicated content in [here](https://github.com/eclipse-tractusx/tractusx-edc/pull/1556/commits/fae5e8bafbc033e16bcfce259fc95cb2ad7b06af).

Expected flow

- Using the new service API, partner adds the participants that wants to obtain the offers from (BPNL's);
- New service stores this data (db or cache);
- Providing the BPNL, the FC obtains the Connectors' URL's from the DS API;
- Using the new service API, FC passes the info from the DS to the new service;
- New service stores this data (db or cache);
- FC, with the ExecutionManager, periodically crawls the participants catalogs based on which ones are stored in the new service;
- FC catalog cache is updated;
- If new participant joins the space, user adds that participant's BPNL through new service API (like first step)

![tnd_proposal](https://github.com/user-attachments/assets/5245f8af-845d-4707-8871-359457d581e3)

⚠️  Any case that the Discovery Service (DS) is used, **a technical user account to access it must be requested**. Without it, the usage of the DS API is limited since a token is used in all relevant requests.

Technically, the new service **would have a simple API** where the user could
- Add a new BPNL
- Remove a BPNL
- List BPNL's (value and connectors associated with it)

> Regarding the service's Persistence Store, any suggestion is welcomed. Having a DB or a local Cache can suffice, but other options like persisting in a file could also be considered. Or a mixed between this options could be the best solution.

Author: Removed the suggestion and focused only on proposed solution. Changed in [here](https://github.com/eclipse-tractusx/tractusx-edc/pull/1556/commits/fae5e8bafbc033e16bcfce259fc95cb2ad7b06af).
